sql: checkpoint a schema change as it progresses

vivekmenezes opened the pull request:
This change writes a checkpoint to the schema change state
within a table descriptor every 10 seconds, so that a node failure
will not restart a schema change from the beginning. A maximum of
10 seconds of work is allowed to get lost.

fixes #7115

will be useful for #7554

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10036)

<!-- Reviewable:end -->

vivekmenezes committed:
sql: checkpoint a schema change as it progresses

This change writes a checkpoint to the schema change state
within a table descriptor every 10 seconds, so that a node failure
will not restart a schema change from the beginning. A maximum of
10 seconds of work is allowed to be lost.

Added a new schema change knob: RunBeforeBackfillChunk which is
a function that gets called before the application of each backfill
chunk.

fixes #7115

will be useful for #7554

paperstreet commented:
Cool, this is a smaller diff than I would have expected. LGTM except for calling everything "checkpointIdx"; that name really confuses me. mutationIdx isn't great, but it's better than checkpointIdx. Maybe we can come up with something nicer than both

Also mention the RunBeforeBackupChunk change in your commit message, please

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/backfill.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO2AtuFDp0bkESSDQd:-KUO2AtvptcGzIcrO01t:bfz2p58) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L161)):*
> ```Go
> 				case *sqlbase.DescriptorMutation_Column:
> 					addedColumnDescs = append(addedColumnDescs, *t.Column)
> 					if columnCheckpointIdx < 0 {
> ```

this is easier to read if you define a const checkpointSentinel = -1 and use an equality check here

---

*[pkg/sql/backfill.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO5VKwfU9B51IYpNGZ:-KUO5VKwfU9B51IYpNG_:b-aapbf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L219)):*
> ```Go
> // getTableSpan returns a span stored at a checkpoint idx, or in the absence
> // of a checkpoint, the span over all keys within a table.
> func (sc *SchemaChanger) getTableSpan(checkpointIdx int) (roachpb.Span, error) {
> ```

what you're calling a checkpoint index is really alwyas an index into the mutations, right? or is that only some uses?

---

*[pkg/sql/backfill.go, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO6Otyk1OOIXh5gGYL:-KUO6Otz27SqFetydzsd:b-pnms31) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L243)):*
> ```Go
> }
> 
> func (sc *SchemaChanger) writeResumeSpan(
> ```

maybeWriteResumeSpan

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-:-KUO7Z1Bjx9Tv5Rk4l2o:boes2pk)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've added RunBeforeBackupChunk to the commit message and renamed checkpointIdx to mutationIdx. I've also added in a further check that it indeed points to a mutation with the correct mutationID

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/backfill.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO2AtuFDp0bkESSDQd:-KUOVpOggTYZv5bWMxkg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L161)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> this is easier to read if you define a const checkpointSentinel = -1 and use an equality check here
</details>
Done.

---

*[pkg/sql/backfill.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO5VKwfU9B51IYpNGZ:-KUOZXzaBcH9s7nY-qiu:b-u3odpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L219)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> what you're calling a checkpoint index is really alwyas an index into the mutations, right? or is that only some uses?
</details>
yeah it's always an index into the mutations.

---

*[pkg/sql/backfill.go, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-KUO6Otyk1OOIXh5gGYL:-KUOVqWFpbLYlFtdgdrD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1c7a2a5f0dfd9b0f5af6b757a1264934727b9c4/pkg/sql/backfill.go#L243)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> maybeWriteResumeSpan
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10036#-:-KUOZejpRgyzAgJ6x1vT:bifukjc)*
<!-- Sent from Reviewable.io -->
